src/sasa/sasa_HASEL.cpp

```diff
@@ -0,0 +1,2716 @@
+  keys.add("compulsory", "NL_STRIDE", "The frequency with which the neighbor list for the calculation of SASA is updated.");
+  keys.add("optional","DELTAGFILE","a file containing the free energy of transfer values for backbone and sidechains atoms. Necessary only if TYPE = TRANSFER. If TYPE = TRANSFER and no DELTAGFILE is provided, the free energy values are those describing the effect of temperature, and are computed using the temperature value passed by the MD engine");
+  keys.add("optional","APPROACH","either approach 2 or 3. Necessary only if TYPE = TRANSFER and no DELTAGFILE is provided. If TYPE = TRANSFER and no DELTAGFILE is provided, the free energy values are those describing the effect of temperature, and the program must know if approach 2 or 3 (as described in Arsiccio and Shea, Protein Cold Denaturation in Implicit Solvent Simulations: A Transfer Free Energy Approach, J. Phys. Chem. B, 2021) needs to be used to compute them");
+}
+
+
+SASA_HASEL::SASA_HASEL(const ActionOptions&ao):
+  PLUMED_COLVAR_INIT(ao),
+  nopbc(false),
+  stride(10),
+  nl_update(0),
+  DeltaGValues("absent"),
+  Ti(0),
+  firstStepFlag(0)
+{
+  rs = 0.14;
+  parse("DELTAGFILE",DeltaGValues);
+  parse("APPROACH", approach);
+  parseAtomList("ATOMS",atoms);
+  if(atoms.size()==0) error("no atoms specified");
+  std::string Type;
+  parse("TYPE",Type);
+  parse("NL_STRIDE", stride);
+  parseFlag("NOPBC",nopbc);
+  checkRead();
+
+  if(Type=="TOTAL") sasa_type=TOTAL;
+  else if(Type=="TRANSFER") sasa_type=TRANSFER;
+  else error("Unknown SASA type");
+
+  switch(sasa_type)
+  {
+  case TOTAL:   log.printf("  TOTAL SASA;"); break;
+  case TRANSFER: log.printf("  TRANSFER MODEL;"); break;
+  }
+
+  log.printf("  atoms involved : ");
+  for(unsigned i=0; i<atoms.size(); ++i) {
+    if(i%25==0) log<<"\n";
+    log.printf("%d ",atoms[i].serial());
+  }
+  log.printf("\n");
+
+  if(nopbc) {
+    log<<"  PBC will be ignored\n";
+  } else {
+    log<<"  broken molecules will be rebuilt assuming atoms are in the proper order\n";
+  }
+
+
+  addValueWithDerivatives(); setNotPeriodic();
+  requestAtoms(atoms);
+
+  natoms = getNumberOfAtoms();
+  AtomResidueName.resize(2);
+  SASAparam.resize(natoms);
+  CONNECTparam.resize(natoms);
+  MaxSurf.resize(natoms);
+  DeltaG.resize(natoms+1);
+  Nlist.resize(natoms);
+
+
+}
+
+
+//splits strings into tokens. Used to read into SASA parameters file and into reference pdb file
+template <class Container>
+void split(const std::string& str, Container& cont)
+{
+  std::istringstream iss(str);
+  std::copy(std::istream_iterator<std::string>(iss),
+            std::istream_iterator<std::string>(),
+            std::back_inserter(cont));
+}
+
+
+//reads input PDB file provided with MOLINFO, and assigns atom and residue names to each atom involved in the CV
+
+void SASA_HASEL::readPDB() {
+  auto* moldat = plumed.getActionSet().selectLatest<GenericMolInfo*>(this);
+  AtomResidueName[0].clear();
+  AtomResidueName[1].clear();
+
```

Comment:
Here moldat could be nullptr. I would suggest to give an error in this case.

Reply:
Thanks a lot for this comment. I added this.